// ngx-material-file-input/bundles/ngx-material-file-input.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/cdk/a11y'), require('@angular/forms'), require('@angular/material'), require('@angular/material/form-field'), require('@angular/cdk/coercion')) :
    typeof define === 'function' && define.amd ? define('ngx-material-file-input', ['exports', '@angular/core', '@angular/cdk/a11y', '@angular/forms', '@angular/material', '@angular/material/form-field', '@angular/cdk/coercion'], factory) :
    (global = global || self, factory(global['ngx-material-file-input'] = {}, global.ng.core, global.ng.cdk.a11y, global.ng.forms, global.ng.material, global.ng.material['form-field'], global.ng.cdk.coercion));
}(this, function (exports, core, a11y, forms, material, formField, coercion) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    /**
     * Optional token to provide custom configuration to the module
     */
    var NGX_MAT_FILE_INPUT_CONFIG = new core.InjectionToken('ngx-mat-file-input.config');
    /**
     * Provide additional configuration to dynamically customize the module injection
     */
    var FileInputConfig = /** @class */ (function () {
        function FileInputConfig() {
        }
        return FileInputConfig;
    }());

    /**
     * The files to be uploaded
     */
    var FileInput = /** @class */ (function () {
        function FileInput(_files, delimiter) {
            if (delimiter === void 0) { delimiter = ', '; }
            this._files = _files;
            this.delimiter = delimiter;
            this._fileNames = (this._files || []).map(function (f) { return f.name; }).join(delimiter);
        }
        Object.defineProperty(FileInput.prototype, "files", {
            get: function () {
                return this._files || [];
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInput.prototype, "fileNames", {
            get: function () {
                return this._fileNames;
            },
            enumerable: true,
            configurable: true
        });
        return FileInput;
    }());

    // Boilerplate for applying mixins to FileInput
    /** @docs-private */
    var FileInputBase = /** @class */ (function () {
        function FileInputBase(_defaultErrorStateMatcher, _parentForm, _parentFormGroup, ngControl) {
            this._defaultErrorStateMatcher = _defaultErrorStateMatcher;
            this._parentForm = _parentForm;
            this._parentFormGroup = _parentFormGroup;
            this.ngControl = ngControl;
        }
        return FileInputBase;
    }());
    /**
     * Allows to use a custom ErrorStateMatcher with the file-input component
     */
    var FileInputMixinBase = material.mixinErrorState(FileInputBase);

    var FileInputComponent = /** @class */ (function (_super) {
        __extends(FileInputComponent, _super);
        /**
         * @see https://angular.io/api/forms/ControlValueAccessor
         */
        function FileInputComponent(fm, _elementRef, _renderer, _defaultErrorStateMatcher, ngControl, _parentForm, _parentFormGroup) {
            var _this = _super.call(this, _defaultErrorStateMatcher, _parentForm, _parentFormGroup, ngControl) || this;
            _this.fm = fm;
            _this._elementRef = _elementRef;
            _this._renderer = _renderer;
            _this._defaultErrorStateMatcher = _defaultErrorStateMatcher;
            _this.ngControl = ngControl;
            _this._parentForm = _parentForm;
            _this._parentFormGroup = _parentFormGroup;
            _this.focused = false;
            _this.controlType = 'file-input';
            _this.autofilled = false;
            _this._required = false;
            _this.accept = null;
            _this.id = "ngx-mat-file-input-" + FileInputComponent_1.nextId++;
            _this.describedBy = '';
            _this._onChange = function (_) { };
            _this._onTouched = function () { };
            if (_this.ngControl != null) {
                _this.ngControl.valueAccessor = _this;
            }
            fm.monitor(_elementRef.nativeElement, true).subscribe(function (origin) {
                _this.focused = !!origin;
                _this.stateChanges.next();
            });
            return _this;
        }
        FileInputComponent_1 = FileInputComponent;
        FileInputComponent.prototype.setDescribedByIds = function (ids) {
            this.describedBy = ids.join(' ');
        };
        Object.defineProperty(FileInputComponent.prototype, "value", {
            get: function () {
                return this.empty ? null : new FileInput(this._elementRef.nativeElement.value || []);
            },
            set: function (fileInput) {
                if (fileInput) {
                    this.writeValue(fileInput);
                    this.stateChanges.next();
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInputComponent.prototype, "placeholder", {
            get: function () {
                return this._placeholder;
            },
            set: function (plh) {
                this._placeholder = plh;
                this.stateChanges.next();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInputComponent.prototype, "empty", {
            /**
             * Whether the current input has files
             */
            get: function () {
                return !this._elementRef.nativeElement.value || this._elementRef.nativeElement.value.length === 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInputComponent.prototype, "shouldLabelFloat", {
            get: function () {
                return this.focused || !this.empty || this.valuePlaceholder !== undefined;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInputComponent.prototype, "required", {
            get: function () {
                return this._required;
            },
            set: function (req) {
                this._required = coercion.coerceBooleanProperty(req);
                this.stateChanges.next();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInputComponent.prototype, "isDisabled", {
            get: function () {
                return this.disabled;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FileInputComponent.prototype, "disabled", {
            get: function () {
                return this._elementRef.nativeElement.disabled;
            },
            set: function (dis) {
                this.setDisabledState(coercion.coerceBooleanProperty(dis));
                this.stateChanges.next();
            },
            enumerable: true,
            configurable: true
        });
        FileInputComponent.prototype.onContainerClick = function (event) {
            if (event.target.tagName.toLowerCase() !== 'input' && !this.disabled) {
                this._elementRef.nativeElement.querySelector('input').focus();
                this.focused = true;
                this.open();
            }
        };
        Object.defineProperty(FileInputComponent.prototype, "fileNames", {
            get: function () {
                return this.value ? this.value.fileNames : this.valuePlaceholder;
            },
            enumerable: true,
            configurable: true
        });
        FileInputComponent.prototype.writeValue = function (obj) {
            this._renderer.setProperty(this._elementRef.nativeElement, 'value', obj instanceof FileInput ? obj.files : null);
        };
        FileInputComponent.prototype.registerOnChange = function (fn) {
            this._onChange = fn;
        };
        FileInputComponent.prototype.registerOnTouched = function (fn) {
            this._onTouched = fn;
        };
        /**
         * Remove all files from the file input component
         * @param [event] optional event that may have triggered the clear action
         */
        FileInputComponent.prototype.clear = function (event) {
            if (event) {
                event.preventDefault();
                event.stopPropagation();
            }
            this.value = new FileInput([]);
            this._elementRef.nativeElement.querySelector('input').value = null;
            this._onChange(this.value);
        };
        FileInputComponent.prototype.change = function (event) {
            var fileList = event.target.files;
            var fileArray = [];
            if (fileList) {
                for (var i = 0; i < fileList.length; i++) {
                    fileArray.push(fileList[i]);
                }
            }
            this.value = new FileInput(fileArray);
            this._onChange(this.value);
        };
        FileInputComponent.prototype.blur = function () {
            this.focused = false;
            this._onTouched();
        };
        FileInputComponent.prototype.setDisabledState = function (isDisabled) {
            this._renderer.setProperty(this._elementRef.nativeElement, 'disabled', isDisabled);
        };
        FileInputComponent.prototype.ngOnInit = function () {
            this.multiple = coercion.coerceBooleanProperty(this.multiple);
        };
        FileInputComponent.prototype.open = function () {
            if (!this.disabled) {
                this._elementRef.nativeElement.querySelector('input').click();
            }
        };
        FileInputComponent.prototype.ngOnDestroy = function () {
            this.stateChanges.complete();
            this.fm.stopMonitoring(this._elementRef.nativeElement);
        };
        FileInputComponent.prototype.ngDoCheck = function () {
            if (this.ngControl) {
                // We need to re-evaluate this on every change detection cycle, because there are some
                // error triggers that we can't subscribe to (e.g. parent form submissions). This means
                // that whatever logic is in here has to be super lean or we risk destroying the performance.
                this.updateErrorState();
            }
        };
        var FileInputComponent_1;
        FileInputComponent.nextId = 0;
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], FileInputComponent.prototype, "autofilled", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], FileInputComponent.prototype, "valuePlaceholder", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], FileInputComponent.prototype, "multiple", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], FileInputComponent.prototype, "accept", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", material.ErrorStateMatcher)
        ], FileInputComponent.prototype, "errorStateMatcher", void 0);
        __decorate([
            core.HostBinding(),
            __metadata("design:type", Object)
        ], FileInputComponent.prototype, "id", void 0);
        __decorate([
            core.HostBinding('attr.aria-describedby'),
            __metadata("design:type", Object)
        ], FileInputComponent.prototype, "describedBy", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", FileInput),
            __metadata("design:paramtypes", [FileInput])
        ], FileInputComponent.prototype, "value", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Object),
            __metadata("design:paramtypes", [Object])
        ], FileInputComponent.prototype, "placeholder", null);
        __decorate([
            core.HostBinding('class.mat-form-field-should-float'),
            __metadata("design:type", Object),
            __metadata("design:paramtypes", [])
        ], FileInputComponent.prototype, "shouldLabelFloat", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean),
            __metadata("design:paramtypes", [Boolean])
        ], FileInputComponent.prototype, "required", null);
        __decorate([
            core.HostBinding('class.file-input-disabled'),
            __metadata("design:type", Object),
            __metadata("design:paramtypes", [])
        ], FileInputComponent.prototype, "isDisabled", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean),
            __metadata("design:paramtypes", [Boolean])
        ], FileInputComponent.prototype, "disabled", null);
        __decorate([
            core.HostListener('change', ['$event']),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", [Event]),
            __metadata("design:returntype", void 0)
        ], FileInputComponent.prototype, "change", null);
        __decorate([
            core.HostListener('focusout'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], FileInputComponent.prototype, "blur", null);
        FileInputComponent = FileInputComponent_1 = __decorate([
            core.Component({
                // tslint:disable-next-line:component-selector
                selector: 'ngx-mat-file-input',
                template: "<input #input type=\"file\" [attr.multiple]=\"multiple? '' : null\" [attr.accept]=\"accept\">\n<span class=\"filename\">{{ fileNames }}</span>\n",
                providers: [{ provide: formField.MatFormFieldControl, useExisting: FileInputComponent_1 }],
                styles: [":host{display:inline-block}:host:not(.file-input-disabled){cursor:pointer}input{width:0;height:0;opacity:0;overflow:hidden;position:absolute;z-index:-1}.filename{display:inline-block}"]
            }),
            __param(4, core.Optional()),
            __param(4, core.Self()),
            __param(5, core.Optional()),
            __param(6, core.Optional()),
            __metadata("design:paramtypes", [a11y.FocusMonitor,
                core.ElementRef,
                core.Renderer2,
                material.ErrorStateMatcher,
                forms.NgControl,
                forms.NgForm,
                forms.FormGroupDirective])
        ], FileInputComponent);
        return FileInputComponent;
    }(FileInputMixinBase));

    var ByteFormatPipe = /** @class */ (function () {
        function ByteFormatPipe(config) {
            this.config = config;
            this.unit = config ? config.sizeUnit : 'Byte';
        }
        ByteFormatPipe.prototype.transform = function (value, args) {
            if (parseInt(value, 10) >= 0) {
                value = this.formatBytes(+value, +args);
            }
            return value;
        };
        ByteFormatPipe.prototype.formatBytes = function (bytes, decimals) {
            if (bytes === 0) {
                return '0 ' + this.unit;
            }
            var B = this.unit.charAt(0);
            var k = 1024;
            var dm = decimals || 2;
            var sizes = [this.unit, 'K' + B, 'M' + B, 'G' + B, 'T' + B, 'P' + B, 'E' + B, 'Z' + B, 'Y' + B];
            var i = Math.floor(Math.log(bytes) / Math.log(k));
            return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
        };
        ByteFormatPipe = __decorate([
            core.Pipe({
                name: 'byteFormat'
            }),
            __param(0, core.Optional()),
            __param(0, core.Inject(NGX_MAT_FILE_INPUT_CONFIG)),
            __metadata("design:paramtypes", [FileInputConfig])
        ], ByteFormatPipe);
        return ByteFormatPipe;
    }());

    var MaterialFileInputModule = /** @class */ (function () {
        function MaterialFileInputModule() {
        }
        MaterialFileInputModule = __decorate([
            core.NgModule({
                declarations: [FileInputComponent, ByteFormatPipe],
                providers: [a11y.FocusMonitor],
                exports: [FileInputComponent, ByteFormatPipe]
            })
        ], MaterialFileInputModule);
        return MaterialFileInputModule;
    }());


    (function (FileValidator) {
        /**
         * Function to control content of files
         *
         * @param bytes max number of bytes allowed
         *
         * @returns
         */
        function maxContentSize(bytes) {
            return function (control) {
                var size = control && control.value ? control.value.files.map(function (f) { return f.size; }).reduce(function (acc, i) { return acc + i; }, 0) : 0;
                var condition = bytes >= size;
                return condition
                    ? null
                    : {
                        maxContentSize: {
                            actualSize: size,
                            maxSize: bytes
                        }
                    };
            };
        }
        FileValidator.maxContentSize = maxContentSize;
    })(exports.FileValidator || (exports.FileValidator = {}));

    exports.ByteFormatPipe = ByteFormatPipe;
    exports.FileInput = FileInput;
    exports.FileInputComponent = FileInputComponent;
    exports.FileInputConfig = FileInputConfig;
    exports.MaterialFileInputModule = MaterialFileInputModule;
    exports.NGX_MAT_FILE_INPUT_CONFIG = NGX_MAT_FILE_INPUT_CONFIG;
    exports.ɵa = FileInputBase;
    exports.ɵb = FileInputMixinBase;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
